AnnoDesigner.Core/DataStructures/QuadTree.cs

```diff
@@ -72,204 +62,235 @@ public Quadrant(Rect extent)
                 {
-                    bottomRight ??= new Quadrant(bottomRightBounds);
-                    childQuadrant = bottomRight;
+                    removed = TopLeft.Remove(item, bounds);
                 }
-                else if (bottomLeftBounds.Contains(bounds))
+                else if (BottomRight != null && bottomRightBounds.Contains(bounds))
                 {
-                    bottomLeft ??= new Quadrant(bottomLeftBounds);
-                    childQuadrant = bottomLeft;
+                    removed = BottomRight.Remove(item, bounds);
                 }
-
-                if (childQuadrant is null)
+                else if (BottomLeft != null && bottomLeftBounds.Contains(bounds))
                 {
-                    ItemsInQuadrant.Add((item, bounds));
+                    removed = BottomLeft.Remove(item, bounds);
                 }
                 else
                 {
-                    childQuadrant.Parent = this;
-                    childQuadrant.Insert(item, bounds);
+                    removed = ItemsInQuadrant.Remove(item);
                 }
 
-                isDirty = true;
+                return removed;
             }
 
             /// <summary>
-            /// Removes an item from this Quadrant
+            /// Clears this collection.
             /// </summary>
-            /// <param name="item"></param>
-            internal void Remove((T item, Rect bounds) item)
+            public void Clear()
             {
-                ItemsInQuadrant.Remove(item);
-                MarkAncestorsAsDirty(); //make sure all ancestors are now marked as requiring an update.
+                TopRight = null;
+                TopLeft = null;
+                BottomRight = null;
+                BottomLeft = null;
+                ItemsInQuadrant.Clear();
             }
 
             /// <summary>
-            /// Marks this quadrant and all parent quadrants as dirty.
+            /// Checks if specified item is inside this collection.
             /// </summary>
-            internal void MarkAncestorsAsDirty()
+            public bool Contains(T item)
             {
-                if (Parent != null)
+                var bounds = item.Bounds;
+                return TopRight != null && TopRight.Extent.IntersectsWith(bounds) && TopRight.Contains(item) ||
+                       TopLeft != null && TopLeft.Extent.IntersectsWith(bounds) && TopLeft.Contains(item) ||
+                       BottomRight != null && BottomRight.Extent.IntersectsWith(bounds) && BottomRight.Contains(item) ||
+                       BottomLeft != null && BottomLeft.Extent.IntersectsWith(bounds) && BottomLeft.Contains(item) ||
+                       ItemsInQuadrant.Contains(item);
+            }
+
+            /// <summary>
+            /// Copies items from this collection to target array.
+            /// </summary>
+            public void CopyTo(T[] array, int arrayIndex)
```

Comment:
Is this method actually used anywhere? It feels like an odd method to have on the quad tree.

Reply:
No, it is there to implement ICollection<T>